src/api/cpp/nixl_v2_1.h

```diff
@@ -0,0 +1,232 @@
+                struct obj::Agent agent;
+            };
+
+            struct AddMemSpace {
+                struct obj::MemSpace           memSpace;
+                uint64_t                       len;
+                std::vector<struct obj::Agent> accessList;
+            };
+
+            struct DelMemSpace {
+                struct obj::MemSpace memSpace;
+            };
+
+            struct Msg {
+                Opcode                 opcode;
+                union {
+                    struct AddAgent    addAgent;
+                    struct DelAgent    delAgent;
+                    struct AddMemSpace addMemSpace;
+                    struct DelMemSpace delMemSpace;
+                };
+            };
+        } // namespace msg
+
+        class Client {
+            public:
+                using RcvMsgCb = std::function<void(const struct Msg &msg)>;
+
+                Client(const struct AgentID &serverID,
+                       const RcvMsgCb       &rcvMsgCb,
+                       const bool           autoPollRcvMsg);
+                ~Client();
+
+                void sendMsg(const struct Msg &msg);
+
+                void pollRcvMsg();
+        };
+    } // namespace pod
+
+    namespace mem {
+        enum class Type {
+            LOCAL_DRAM,
+            LOCAL_VRAM,
+            LOCAL_BLK,
+            LOCAL_OBJ,
+            LOCAL_FILE,
+            REMOTE_AGENT
+        };
+
+        class Space {
+        public:
+            struct Local {
+                Type     type;
+                int      dev_id;
+                uint64_t len;
+            };
+
+            Space(const struct Local &local);
+            Space(const struct pod::obj::MemSpace &podMemSpace);
+
+            // generate a key for granting access to a local mem space for remote agents
+            uint32_t genPodMemSpaceKeyForLocalMemSpace();
+
+            ~Space();
+        };
+    } // namespace mem
+ 
+    struct MemXferType {
+        mem::Type dst;
+        mem::Type src;
+    };
+
+    namespace container {
+        using BackendType = std::string;
+
+        struct BackendTypeParams {
+            std::vector<MemXferType> supportedMemXferTypes;
+        };
+
+        void getAvailBackendTypes(std::vector<BackendType> &backendTypes);
+
+        void getBackendTypeParams(const struct BackendTypeParams &params);
+    }; // namespace container
+
+    class Agent {
+        enum class PostState {
+            NOT_POSTED,
+            POSTED,
+            COMPLETED
+        };
+
+        enum class PostCompletionStatus {
+            OK,
+            ERROR
+        };
+
+        class MemXfer {
+        public:
+            struct MemXferBuf {
+                class mem::Space space;
+                size_t           ofst;
+                size_t           len;
+            };
+
+            PostState            postState;
+            PostCompletionStatus postCompletionStatus;
+
+            // dst and src must have the same length
```

Comment:
If these two collections must have the same size, then it should be a single collection containing both src and dst.